feat: introduce `libp2p-connection-limits` connection management module

thomaseizinger opened the pull request:
## Description

<!-- Please write a summary of your changes and why you made them.-->
<!-- This section will appear as the commit message after merging. Please craft it accordingly. -->

This patch deprecates the existing connection limits within `Swarm` and uses the new `NetworkBehaviour` APIs to implement it as a plugin instead.

Related #2824.

## Notes

~This is a WIP in that it still needs tests but overall, this is pretty much what it will look like.~

<!-- Any notes or remarks you'd like to make about the PR. -->

## Links to any relevant issues

<!-- Reference any related issues.-->

## Open Questions

<!-- Unresolved questions, if any. -->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

thomaseizinger committed:
Use GATs to avoid lifetimes on trait interface

thomaseizinger committed:
Bump MSRV accordingly

thomaseizinger committed:
Bump MSRV of libp2p-metrics

thomaseizinger committed:
Merge branch 'master' into feat-kademlia/use-gats

thomaseizinger committed:
Refactor dialling to have a more linear flow

thomaseizinger committed:
Introduce new callbacks for connection management

thomaseizinger committed:
Fix up some more code

thomaseizinger committed:
Complete deprecation notice

thomaseizinger committed:
Fix some more errors

thomaseizinger committed:
Refine deprecation message

thomaseizinger committed:
Delegate to `addresses_of_peer` by default

thomaseizinger committed:
Add hack to clear listen addresses

thomaseizinger committed:
Merge branch 'master' into feat-kademlia/use-gats

thomaseizinger committed:
Merge branch 'feat-kademlia/use-gats' of github.com:libp2p/rust-libp2p into feat-kademlia/use-gats

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Use an atomic counter to avoid collisions on connection IDs

thomaseizinger committed:
Fix minor compile error

thomaseizinger committed:
Fill in a few more bits

thomaseizinger committed:
Don't be generic over transport

thomaseizinger committed:
Remove type parameter for transport error

thomaseizinger committed:
Remove unnecessary type hint

thomaseizinger committed:
Remove type parameters from `ConcurrentDial`

thomaseizinger committed:
Fix rustfmt

thomaseizinger committed:
Merge branch 'feat-kademlia/use-gats' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch '2824-simplify-dial-opts' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch '2824-no-generic-transport-pool' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Remove `PendingConnectionDenied`

This is handled by `DialError` and `ListenFailure`.

thomaseizinger committed:
Handle new DialError variants

thomaseizinger committed:
Make `libp2p-kad` compile

thomaseizinger committed:
Fix a few more errors

thomaseizinger committed:
No more compile errors!

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Undo explicit `Sized` bound

thomaseizinger committed:
Let user know that we are discarding addresses

thomaseizinger committed:
Fix unused variable warning

thomaseizinger committed:
Fix unused import warnings

thomaseizinger committed:
Port `dcutr` protocol

thomaseizinger committed:
Re-implement relay

thomaseizinger committed:
Fix some deprecation warnings

thomaseizinger committed:
Fix some deprecation warnings kad

thomaseizinger committed:
Port autonat

thomaseizinger committed:
Port identify

thomaseizinger committed:
Fix more deprecation warnings

thomaseizinger committed:
Ignore clippy warning

thomaseizinger committed:
Fix derive and remove todo

thomaseizinger committed:
Make derive tests compile by using type alias

thomaseizinger committed:
Ignore clippy warnings in generated code

thomaseizinger committed:
Fix more warnings by allowing deprecations

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Remove unnecessary `extern crate`

thomaseizinger committed:
Revert let-else usage

thomaseizinger committed:
Less let else

thomaseizinger committed:
No let else in mdns

thomaseizinger committed:
No let-else in dcutr

thomaseizinger committed:
No let else in rendezvous

thomaseizinger committed:
Last let-else usage

thomaseizinger committed:
Remove fn from bad merge

thomaseizinger committed:
Minimize diff

thomaseizinger committed:
Attempt to minimize diff

thomaseizinger committed:
Undo re-ordering to reduce diff

thomaseizinger committed:
Minimize diff (and fix bad merge)

thomaseizinger committed:
Remove default impl for `ConnectionId` in favor of `next`

thomaseizinger committed:
Deprecate `IntoMultiHandler`

thomaseizinger committed:
Implement dial failure reporting

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Use single quotes to avoid running commands

thomaseizinger committed:
Use intermediary variable

thomaseizinger committed:
Make connection IDs globally unique

thomaseizinger committed:
Deprecate old new function

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Merge branch 'master' into no-run-title-as-command

thomaseizinger committed:
Remove `handler` field from `NetworkBehaviourAction::Dial`

thomaseizinger committed:
Port dcutr protocol

thomaseizinger committed:
Port relay protocol

thomaseizinger committed:
Add changelog entry to swarm

thomaseizinger committed:
Merge branch 'no-run-title-as-command' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Remove ordering functionality from `ConnectionId`

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' of github.com:libp2p/rust-libp2p into 2824-unique-connection-ids

thomaseizinger committed:
Remove unnecessary visibility qualifiers

thomaseizinger committed:
Convert `DialOpts` to struct variant

thomaseizinger committed:
Eagerly set defaults in `DialOpts`

Instead of carrying around an enum and parsing its contents later,
directly set the correct values as we construct the final `DialOpts`.

thomaseizinger committed:
Merge branch '2824-refactor-dial-opts' into 2824-no-handler-in-dial

thomaseizinger committed:
Make `ConnectionId` part of `DialOpts`

thomaseizinger committed:
Inline `_dial`

thomaseizinger committed:
Rustfmt

thomaseizinger committed:
Fix wrong default

thomaseizinger committed:
Merge branch '2824-refactor-dial-opts' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' of github.com:libp2p/rust-libp2p into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Expand changelog entry

thomaseizinger committed:
Move most important changelog entry to the top

thomaseizinger committed:
Add more text to changelog entry

thomaseizinger committed:
Fix formatting

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Fix most errors

mergify[bot] committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch 'master' into 2824-unique-connection-ids

thomaseizinger committed:
Merge branch '2824-unique-connection-ids' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Fix implementation of dcutr

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Undo accidential change

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Attempt to reduce diff

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Remove unnecessary `..`

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
WIP

thomaseizinger committed:
Remove unused error variant

thomaseizinger committed:
Apply suggestions from code review

thomaseizinger committed:
Merge branch '2824-remove-unused-error' into 2824-introduce-listen-error

thomaseizinger committed:
WIP

thomaseizinger committed:
Remove unused `DialError::ConnectionIo` variant

thomaseizinger committed:
Merge branch '2824-remove-unused-error' into 2824-remove-unused-connection-error

thomaseizinger committed:
add changelog entry

thomaseizinger committed:
Fix missing colon

thomaseizinger committed:
Merge branch '2824-remove-unused-connection-error' into 2824-introduce-listen-error

thomaseizinger committed:
Remove unused `ConnectionIo` variant from `ListenError`

thomaseizinger committed:
Update docs

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Update changelog

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Merge branch '2824-introduce-listen-error' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Report cause of denied connection

thomaseizinger committed:
Introduce dedicated error type

thomaseizinger committed:
Don't comment docs

thomaseizinger committed:
Fix docs

thomaseizinger committed:
Migrate `CallTraceBehaviour`

thomaseizinger committed:
Temporarily allow usage of `new_handler`

thomaseizinger committed:
Refactor gossipsub config handler to avoid calls to trait functions

thomaseizinger committed:
Undo bad doc comment change

thomaseizinger committed:
Fix rustdocs

thomaseizinger committed:
Fix deprecation warnings

thomaseizinger committed:
Fix failing gossipsub test

thomaseizinger committed:
Fix more docs

thomaseizinger committed:
Fix beta clippy

thomaseizinger committed:
Actually fix beta clippy

thomaseizinger committed:
Resolve remaining todos

thomaseizinger committed:
Implement connections limits using `NetworkBehaviour`

thomaseizinger committed:
Always remove connections from pending after successful upgrade

thomaseizinger committed:
Assert that handler never sends events

mxinden commented on swarm/src/connection_limits.rs:
```diff
@@ -0,0 +1,196 @@
+use crate::{
```
License header missing :angel:

mxinden reviewed (commented):
:sparkles: 

Ignoring the fact that this is still a draft pull request, change looks good to me. Just missing changelog entries.

thomaseizinger committed:
Don't use else block if we return in the previous one

thomaseizinger committed:
Remove closure in favor of spawning connection from the outside

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Add `ConnectionId` to `ListenFailure`

thomaseizinger committed:
Clean up pending commands on failure

thomaseizinger committed:
Add docs and fix bug

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Consistently use `THandlerInEvent`

thomaseizinger committed:
Use type alias

thomaseizinger committed:
Remove unnecessary indentation and "quote"-ing

thomaseizinger committed:
Fix bad changelog entry

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Spawn connection at right point in time

thomaseizinger committed:
Attempt to reduce diff

thomaseizinger committed:
Reduce diff further

thomaseizinger committed:
Less diff ...

thomaseizinger committed:
Don't override deprecated callbacks anywhere

thomaseizinger committed:
Better order of callbacks

thomaseizinger committed:
Spawn connection at right point in time

Otherwise the num-established counter is wrong.

thomaseizinger committed:
Be consistent in order of parameters

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Reduce diff

thomaseizinger committed:
Fix rustdoc

thomaseizinger committed:
One more missing rustdoc link

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Remove `banned_peer_connections`

We no longer add connections for banned peers to the pool so we don't
have to filter them later.

thomaseizinger committed:
Fix test

Now that we no longer add connections to banned peers, we need to:

- Reduce the number of expected connections by 1 (i.e. delete line 1954)
- Act in `Stage::BannedDial` earlier (i.e. remove the `if`)

thomaseizinger committed:
Remove duplication around detecting relayed addresses in `libp2p-relay`

thomaseizinger committed:
Extract utility fn for dcutr protocol

thomaseizinger committed:
Remove unused import

thomaseizinger committed:
Remove underscore from used variable

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Add changelog entry

thomaseizinger committed:
Add footnote

thomaseizinger committed:
Don't use footnotes

thomaseizinger committed:
Merge branch 'master' into 2824-no-handler-in-dial

thomaseizinger committed:
Inline `spawn` function

thomaseizinger committed:
Box future inside `spawn`

thomaseizinger committed:
Merge impl blocks

thomaseizinger committed:
Gracefully close connection to banned peer

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Ensure new connections are either spawned or gracefully closed

thomaseizinger committed:
Ignore error when closing connection

thomaseizinger committed:
Merge branch '2824-no-handler-in-dial' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Fix compile error

thomaseizinger committed:
Restore `IntoEitherHandler`

thomaseizinger committed:
Don't underscore used variable

thomaseizinger committed:
Merge branch 'master' into 2824-deprecate-into-connection-handler

thomaseizinger committed:
Undo merge conflicts with #3465

thomaseizinger committed:
Call out limitations in Rust compiler

thomaseizinger committed:
Move manage conn entry to top of changelog

This one is more important.

thomaseizinger committed:
Fix fmt

thomaseizinger committed:
Merge branch '2824-deprecate-into-connection-handler' into 2824-connection-limits

thomaseizinger committed:
Add license header

thomaseizinger committed:
Fix compile errors

thomaseizinger committed:
Port tests

thomaseizinger committed:
Restore old tests and deprecate more things

thomaseizinger committed:
Merge branch 'master' into 2824-connection-limits

thomaseizinger committed:
Add changelog entry and deprecation

thomaseizinger committed:
Merge branch 'master' into 2824-connection-limits

thomaseizinger committed:
Update changelog

thomaseizinger committed:
Bump swarm version

mxinden commented on swarm/src/connection_limits.rs:
```diff
@@ -0,0 +1,422 @@
+use std::task::{Context, Poll};
+use void::Void;
+
+pub struct Behaviour {
```
The deprecation note on the old connection limiting will point to this type. What do you think of including a small example showcasing how one can use this `NetworkBehaviour`?

thomaseizinger replied:
Done!

mxinden reviewed (commented):
:rocket: great to see the new feature in action. One comment, otherwise looks good to me.

thomaseizinger committed:
Merge branch 'master' into 2824-connection-limits

thomaseizinger committed:
Add docs to `connection_limits::Behaviour`

mxinden commented on swarm/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+# 0.42.1 [unreleased]
```
```suggestion
# 0.42.1
```

I can cut a new release of `libp2p-swarm` directly once merged. What do you think?

thomaseizinger replied:
I'd want to replace the banning API as well do make #2824 "complete" for now.

mxinden reviewed (approved):
Ready to merge from my end. Let me know in case you want a new release of `libp2p-swarm` right away @thomaseizinger.

thomaseizinger committed:
Merge branch 'master' into 2824-connection-limits

thomaseizinger committed:
Reorganise code a bit

We deprecate the old error and use the opportunity to change the
shape of the new `Exceeded` error.

thomaseizinger committed:
Fix doc comments

thomaseizinger committed:
Merge branch 'master' into 2824-connection-limits

thomaseizinger committed:
Move connection limits into separate crate

thomaseizinger committed:
Unify deprecation warnings

thomaseizinger committed:
Update lockfile

thomaseizinger committed:
Fix rustdoc links

thomaseizinger committed:
Add deprecation since version

thomaseizinger committed:
Fix doc tests

thomaseizinger committed:
Undo breaking change

Thanks `cargo semver-checks`!

thomaseizinger committed:
Remove duplicated import

thomaseizinger committed:
Merge branch 'master' into 2824-connection-limits

thomaseizinger committed:
Don't use deprecated APIs

mxinden commented on misc/connection-limits/src/lib.rs:
```diff
@@ -0,0 +1,483 @@
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
```
We overload the string _protocol_ with many meanings. The main one for in relation with the `protocols/` folder being that all things in their implement `NetworkBehaviour`.

Given that `libp2p_connection_limits::Behaviour` implements `NetworkBehaviour` I think it should live in `protocols/`.

thomaseizinger replied:
> The main one for in relation with the `protocols/` folder being that all things in their implement `NetworkBehaviour`.

We already have multiple utilities in `libp2p-swarm` that implement `NetworkBehaviour` without living under `protocols/`. To me, a libp2p "protocol" must have some form of network communication. This one here uses the `dummy::ConnectionHandler` so it will never perform any IO, hence I placed it in `misc/`.

thomaseizinger replied:
As you know, I am not a massive fan of the directory structure anyway but putting this one into `protocols/` actually feels wrong.

Protocols should be present on the wire, this one here is just an internal `rust-libp2p` concept.

`misc/` is not a great name either because everything fits the description :)

With `libp2p-identity` merged, I'd like to push for the next step towards moving around the crate boundaries. Once we have all that in place, I think it makes sense to think about new directories.

Thoughts?

mxinden commented on misc/connection-limits/CHANGELOG.md:
```diff
@@ -0,0 +1,3 @@
+# 0.1.0 - unreleased
```
Nit pick, we recommend the below. I don't care which syntax we use, as long as we stay consistent. Makes my life creating the `docs(CHANGELOG): Prepare xxx` a bit easier.

```suggestion
# 0.1.0 [unreleased]
```

https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases

thomaseizinger replied:
`[unreleased]` is a link reference that doesn't point to anything so I stopped using that. Can update the docs if you agree.

mxinden replied:
Never thought about that. Yes, patch would be appreciated.

thomaseizinger replied:
PR here: https://github.com/libp2p/rust-libp2p/pull/3615

mxinden reviewed (commented):
Looks good to me apart from suggestion to move to `protocols/`.

thomaseizinger committed:
Add changelog entry and bump version

thomaseizinger committed:
Update Cargo.lock

mergify[bot] committed:
Merge branch 'master' into 2824-connection-limits
